File: SAUCEDEMO-Playwright/TypeScript/resources/PageObjects/SAUCEDEMO/productPage.ts
import { BasePage } from "../../basePage";

export class SDProductPage extends BasePage{
    productPageLocators = {
        productHeader: "xpath=//span[@class='title' and text()='Products']",
        itemBox: {
            mainBox: "xpath=(//div[@class='inventory_item' and .//div[@class='inventory_item_name ' and contains(text(),'[TO_CHANGE]')]])[1]",
            itemImg: "/div[@class='inventory_item_img']", // use after mainBox
            itemName: "//div[@class='inventory_item_name ']",
            itemDesc: "//div[@class='inventory_item_desc']",
            itemPrice: "//div[@class='inventory_item_price']",
            addBtn: "//button",
        },
        filter: "xpath=//select[@class='product_sort_container']"
    }

    productsDetail: Record<string, any> = {};



    /**
     * Add or remove product into the cart based on given name, 
     * also check that item really added or not.
     * No error if items is not visible on page nor available.
     * @param products  products as array. (e.g., "Bike Light", "Fleeces")
     * @param [isAdd=true] add item if true, remove if false, default is true
     * @returns none.
     */

    async addOrRemoveProducts(products: string[], isAdd: boolean = true){
        for (let product of products){
            let productBox = this.productPageLocators.itemBox['mainBox'].replace('[TO_CHANGE]',product)
            let addBtn = productBox + this.productPageLocators.itemBox['addBtn']

            let btnText
            // can't find other way to handle failed from timeout.
            try {
                btnText = await this.page.locator(addBtn).textContent({ timeout: 2000 })
            } catch {
                btnText = null
            }

            if (isAdd && btnText == 'Add to cart') {
                await this.page.locator(productBox + this.productPageLocators.itemBox['addBtn']).click()
            } else if (!isAdd && btnText == 'Remove') {
                await this.page.locator(productBox + this.productPageLocators.itemBox['addBtn']).click()
            } else {
                console.log(`\nThere is no product named : ${product} or product is already added or removed.`);
                console.log(`Method want to delete : ${isAdd}, current btn status : ${btnText}\n`)
            }

            if (btnText !== null) {
                btnText = await this.page.locator(addBtn).textContent()
                let verify = (isAdd) ? 'Remove' : 'Add to cart'
                await this.page.waitForTimeout(1000)   // wait for button text to change first.
                await this.expect(this.page.locator(addBtn)).toHaveText(verify, { ignoreCase: true })
            }
            }
    
    }

    /**
     * Get products details based on given name,
     * No error if items is not visible on page nor available.
     * @param products  products as array. (e.g., "Bike Light", "Fleeces")
     * @returns Object
     */
    async getProducts(products: string[]) {

        for (const product of products) {
            let product_box = this.productPageLocators.itemBox['mainBox'].replace('[TO_CHANGE]', product)

            try {
                let productDesc = await this.page.locator(product_box + this.productPageLocators.itemBox['itemDesc']).textContent({ timeout: 2000 })
                let productPrice = await this.page.locator(product_box + this.productPageLocators.itemBox['itemPrice']).textContent({ timeout: 2000 })
                this.productsDetail[product] = { "description": productDesc, "price": productPrice }

            } catch {
                console.log(`There is no product named : ${product}`);
            }

        }
        console.log(this.productsDetail);
        return this.productsDetail
    }

    /**
     * select filter by its values (force to use value only)
     * @param method  the value to be select. (e.g., "za", "lohi")
     */
    async changeFilterByValue(method: string = 'az') {
        const methods = ['az', 'za', 'lohi', 'hilo']
        const filterLocator = this.page.locator(this.productPageLocators.filter)
        if (methods.includes(method)) {
            await filterLocator.selectOption({ value: method })
            await this.expect(filterLocator).toHaveValue(method)
        } else {
            console.log(`There is no such ${method} in ${methods}`);
        }
    }
    
}

File: SAUCEDEMO-Playwright/Python/resources/PageObjects/SAUCEDEMO/product_page.py
from resources.PageObjects.SAUCEDEMO.sauce_common import SauceDemoBase



class ProductPage(SauceDemoBase):

    PRODUCT_PAGE_LOCATORS = {
        "productHeader": "xpath=//span[@class='title' and text()='Products']",
        "itemBox": {
            "mainBox": "xpath=(//div[@class='inventory_item' and .//div[@class='inventory_item_name ' and contains(text(),'[TO_CHANGE]')]])[1]",
            "itemImg": "/div[@class='inventory_item_img']",  # use after mainBox
            "itemName": "//div[@class='inventory_item_name ']",
            "itemDesc": "//div[@class='inventory_item_desc']",
            "itemPrice": "//div[@class='inventory_item_price']",
            "addBtn": "//button",
        },
        "filter": "xpath=//select[@class='product_sort_container']",
        "cart_badge" : "xpath=//span[@class='shopping_cart_badge']"
    }

    products_detail = {}


    def add_or_remove_products(self, products: list):
        """Add specific products to cart from the product page.

        Clicks the add button for each product in the list.
        
        Args:
            products (list): List of product names to add to cart.

        """
        for product in products:
            product_box = self.PRODUCT_PAGE_LOCATORS["itemBox"]["mainBox"].replace(
                "[TO_CHANGE]", product
            )
            add_btn = product_box + self.PRODUCT_PAGE_LOCATORS["itemBox"]["addBtn"]

            self.page.wait_for_timeout(1000)

            if self.page.locator(add_btn).is_visible():
                self.page.locator(add_btn).click()
            else:
                print(f"\nThere is no product named : {product}")


    def get_products_detail(self, products: list):
        """Extract description and price for specific products.

        Retrieves product details for each product in the list and stores them internally.

        Args:
            products (list): List of product names to get details for.

        Returns:
            dict: Dictionary with product names as keys and product details as values.
                Structure: {'product_name': {'description': str, 'price': str}, ...}

        Example:
            >>> products = ['Sauce Labs Backpack', 'Sauce Labs Bike Light']
            >>> products_detail = get_products_detail(products)
            {'Sauce Labs Backpack': {'description': '...', 'price': '$29.99'}, ...}
        """
        for product in products:
            product_box = self.PRODUCT_PAGE_LOCATORS["itemBox"]["mainBox"].replace(
                "[TO_CHANGE]", product
            )

            try:
                product_desc = self.page.locator(
                    product_box + self.PRODUCT_PAGE_LOCATORS["itemBox"]["itemDesc"]
                ).text_content()
                product_price = self.page.locator(
                    product_box + self.PRODUCT_PAGE_LOCATORS["itemBox"]["itemPrice"]
                ).text_content()
                self.products_detail[product] = {
                    "description": product_desc,
                    "price": product_price,
                }
            except:
                print(f"\nThere is no product named : {product}")
        
        print(self.products_detail)
        return self.products_detail

    def change_filter_by_value(self, method="az"):
        """Change product sorting filter by selected method.

        Args:
            method (str): Sorting method. Valid options: 'az' (A-Z), 'za' (Z-A), 
                         'lohi' (low to high price), 'hilo' (high to low price).
                         Defaults to 'az'.

        Raises:
            Exception: If the specified method is not in the valid options list.
        """
        methods = ["az", "za", "lohi", "hilo"]
        
        if method not in methods :
            raise Exception(f"There is no such {method} in {methods}")
        
        filter_locator = self.page.locator(self.PRODUCT_PAGE_LOCATORS["filter"])
        filter_locator.select_option(value=method)
        self.expect(filter_locator,f'locator suppose to have value of {method}').to_have_value(method)
        
    def verify_product_page(self):
        """Verify that the product page has loaded successfully."""
        self.verify_page_arrive(self.PRODUCT_PAGE_LOCATORS["productHeader"])
